lib/datastore/datastore_container.es6.js

```diff
@@ -240,6 +152,139 @@ foam.CLASS({
         });
       },
     },
+  ],
+});
+
+foam.ENUM({
+  name: 'DatastoreContainerMode',
+  package: 'org.chromium.apis.web',
+
+  requires: [
+    'foam.box.BoxRegistry',
+    'foam.box.SkeletonBox',
+    'foam.box.node.ForkBox',
+    'foam.core.StubFactorySingleton',
+    'foam.dao.MDAO',
+    'foam.dao.ReadOnlyDAO',
+    'foam.dao.SyncDAO',
+    'foam.dao.VersionNoDAO',
+    'foam.dao.sync.VersionedSyncRecord',
+    'org.chromium.apis.web.ConfluenceClientDAO',
+    'org.chromium.apis.web.ConfluenceSyncDAO',
+  ],
+  imports: ['info'],
+
+  properties: [
+    {
+      class: 'Function',
+      name: 'getServiceDAO',
+    },
+    {
+      name: 'stubFactory',
+      factory: function() {
+        return this.StubFactorySingleton.create();
+      },
+    },
+  ],
+
+  methods: [
+    function getForkNodeParams_() {
+      return ['--max_old_space_size=4000'];
+    },
+    function getForkScriptPath_() {
+      return require('path').resolve(`${__dirname}/../../main/forkScript.js`);
+    },
+  ],
+
+  values: [
+    {
+      name: 'WEB_SERVICE',
+      documentation: `Initialize DAOs for exposing via a web service. The web
+          service provides read-only access to Datastore, synchronizes a local
+          cache, then redirects requests directly to the local cache.`,
+      getServiceDAO: function(versionedCls, baseRemoteDAO, ctx) {
+        const proxy = this.ReadOnlyDAO.create({
+          delegate: baseRemoteDAO,
+        }, ctx);
+        const sync = this.ConfluenceClientDAO.create({
+          // Create fork, and stub its BoxRegistry.
+          delegate: this.stubFactory.get(this.BoxRegistry).create({
+            delegate: this.ForkBox.create({
+              nodeParams: this.getForkNodeParams_(),
+              childScriptPath: this.getForkScriptPath_(),
+            }, ctx),
+          // Register skeleton over ConfluenceSyncDAO. ConfluenceClientDAO's
+          // delegate is box returned from register() call.
+          }, ctx).register(null, null, this.SkeletonBox.create({
+            data: this.ConfluenceSyncDAO.create({
+              of: versionedCls,
+              remoteDAO: baseRemoteDAO,
+            })
+          })),
+        }, ctx);
+
+        this.info(`Serving read-only DAO of ${versionedCls.id} from Datastore`);
+        sync.hasSynced().then(function() {
+          proxy.delegate = sync;
```

Comment:
Is this potentially racy, or does the mostly single-threaded nature of JS save you from having this switched out while something is using it?

Reply:
This should be fine. In theory, a `ProxyDAO` implementation could asynchronously access its delegate twice under the assumption that it hasn't changed, but that would be a bug in the proxy's implementation. `ReadOnlyDAO` does not do this.